### ISSUE TITLE
generate a weighted straight skeleton from files? 

### ISSUE THREAD

**Issue by pcace** [CGAL/cgal#7612]
Hi there, 

this question is more general cgal realated.  before i start i have to say that i have very limited knowledge about c++. i am trying to get a simple little program to run wich generates a weighted straight skeleton from inputs and returns vertices and indices of the resulting extruded skelton. 

i compiled the code from  test_sls_weighted_polygons_with_holes to just get to know the algorithm. the thing i cannot find is if there is (and i guess there is) any kind of type wich would be used to read the desired polygon + holes  + weights from a text file as an argument.

there are multiple *.poly files in the test/data folder. so i guess it should be possible to load polygons from a file. but how would it work with weights? what i want to achieve in the end is something like: `./Create_weighted_straight_skeleton_mesh <polygonlist.poly> <holes.poly>  <weights.noIdea>`

so in short i guess the question is how can i load weights for a polygon from a file?

thank you so much!

**Comment by MaelRL** [comment]
You can have a look at this example: https://github.com/CGAL/cgal/blob/master/Straight_skeleton_2/examples/Straight_skeleton_2/extrude_skeleton.cpp. It reads polygons and weights from files and computes the extrusion of the skeleton. If you only need the skeleton, you can use the functions from ` test_sls_weighted_polygons_with_holes.cpp`.

The polygon with holes files are in `*.poly` files, the idea is that you have the outer boundary first, with vertices in counterclockwise order, and then the inner holes, with vertices in clockwise order. For weights, there is no specific file format. In this example, I have put them in `*.dat` file, with the same order as for the polygons: one weight per line. See the help with `./extrude_skeleton -h`.

The code that actually performs the read is in: https://github.com/CGAL/cgal/blob/master/Straight_skeleton_2/examples/Straight_skeleton_2/include/CGAL/input_helpers.h.

Note that the doc for 5.6 is there: https://doc.cgal.org/5.6-beta1/Straight_skeleton_2/index.html.

**Comment by pcace** [comment]
Hi, 

thanks a lot for helping out so fast! 
a little thing first: this example seems to not accept *.poly files for -i, at least it throws  this error when using a *.poly file: 

```
Error: unknown file extension: poly
Error: failure during polygon read
```
using *.dat instead seems to work a little better using these two files: 

`poly.dat`:
```
1
5
0 0
8 0
8 1
8 2
0 2
```
and `weights.dat`: 
`1
1
1
1
1`

but still i think i dont quite get it, because i get this error: 

```
./extrude_skeleton -i poly.dat -w weight.dat 
Warning: input polygon not simple (hopefully it is strictly simple...)
Polygon with border of size: 4
Reading input(s) took 0.00031685829162597656 s.
terminate called after throwing an instance of 'CGAL::Precondition_exception'
  what():  CGAL ERROR: precondition violation!
Expr: std::distance(outer_contour_vertices_begin, outer_contour_vertices_end) == std::distance(outer_contour_weights_begin, outer_contour_weights_end)
File: /home/user/git/cgal/Straight_skeleton_2/include/CGAL/create_weighted_straight_skeleton_2.h
Line: 57
Abgebrochen (Speicherabzug geschrieben)
```

any idea what i am doing wrong here? 

thanks a lot again!

**Comment by MaelRL** [comment]
Indeed, the input_helpers.h only supports .dat in this version, sorry.

For weights, only one value per line.

**Comment by pcace** [comment]
Hi, 

sorry - github seems to mess up the weights comment... it is one value per line. 

![Bildschirmfoto vom 2023-07-20 15-13-53](https://github.com/CGAL/cgal/assets/1011875/124ae624-a25b-47c9-acfa-eb4e809b15d7)

**Comment by MaelRL** [comment]
In the .dat format, you don't have the number of polygons at the top.

See for example a rectangle with two holes:
```
4
0 0
10 0
10 10
0 10

4
6 3
7 2.5
8 1
4 1

4
4 4
5 4
5 5.5
4 5.5
```

**Comment by pcace** [comment]
Oh!! Thank you so much!! i thought i need the numbering of each polygon as well. 
Thank you again for your help it works now just perfectly!

**Comment by afabri** [comment]
As it is a new feature it would be cool to learn what you use it for.

**Comment by pcace** [comment]
hi, its a very early idea if it would be possible to use its resulting volume and height  for calculating distance space between complex buildings. i am not quite sure if this is a good aproach, but maybe ;) 
And since there are parts of buildings with different rules for the distance i think using the weight feature could
![Download](https://github.com/CGAL/cgal/assets/1011875/0d2a01a4-96af-46e3-906c-4fd95961a85f)
 be the key.

**Comment by MaelRL** [comment]
> hi, its a very early idea if it would be possible to use its resulting volume and height for calculating distance space (minimal distance where its not allowed to build) between complex buildings. i am not quite sure if this is a good aproach, but maybe ;) And since there are parts of buildings with different rules for the distance i think using the weight feature could be the key. ![Download](https://user-images.githubusercontent.com/1011875/254917108-0d2a01a4-96af-46e3-906c-4fd95961a85f.jpeg)

Very interesting! Note that you also have an API that uses angles, if that is more intuitive than weights in your context.

It is also some very recent code; if you see something that can be improved, or a feature that seems to be missing, don't hesitate to get back to us :).

**Comment by pcace** [comment]
Thank you very much! And i will get back to you if i find something!

**Comment by pcace** [comment]
> > hi, its a very early idea if it would be possible to use its resulting volume and height for calculating distance space (minimal distance where its not allowed to build) between complex buildings. i am not quite sure if this is a good aproach, but maybe ;) And since there are parts of buildings with different rules for the distance i think using the weight feature could be the key. ![Download](https://user-images.githubusercontent.com/1011875/254917108-0d2a01a4-96af-46e3-906c-4fd95961a85f.jpeg)
> 
> Very interesting! Note that you also have an API that uses angles, if that is more intuitive than weights in your context.
> 
> It is also some very recent code; if you see something that can be improved, or a feature that seems to be missing, don't hesitate to get back to us :).

Hi, i am not sure if i should open up another issue, but i think i have something wich might be changed: 
generating an extruded skeleton (see extrude_skeleton.cpp) with weight 0 is possible.

when using create_interior_weighted_skeleton_and_offset_polygons_with_holes_2 it fails when there is a 0 in the weights. 

it would be very nice to get the resulting offsets even when there is one section with weight 0.

here is the output for a polygon with two holes and weights >0

```
Weights:
1 1 1 1 
1 0.20000000000000001 1 0.20000000000000001 
1 1 1 1 
Polygon_with_holes list with 1 element
Polygon_with_holes having 2 holes
Polygon with 4 vertices
(0.10000000000000001,0.10000000000000001)
(9.9000000000000004,0.10000000000000001)
(9.9000000000000004,9.9000000000000004)
(0.10000000000000001,9.9000000000000004)
Polygon with 4 vertices
(6.2783970920917591,3.4198184483290688)
(6.6212605693982534,3.2483867096758208)
(8.4535183758487999,0.50000000000000011)
(3.3585786437626908,0.50000000000000011)
Polygon with 4 vertices
(3.9000000000000004,5.6000000000000005)
(5.1000000000000005,5.6000000000000005)
(5.1000000000000005,3.9000000000000004)
(3.9000000000000004,3.9000000000000004)
```
here is the error when having a 0 in the weights file: 


```
Weights:
1 1 1 1 
1 0 1 0 
1 1 1 1 
terminate called after throwing an instance of 'CGAL::Precondition_exception'
  what():  CGAL ERROR: precondition violation!
Expr: ss.get() != 0
File: /home/johannes/git/cgal/Straight_skeleton_2/include/CGAL/create_offset_polygons_2.h
Line: 187
Abgebrochen (Speicherabzug geschrieben)
```

since i am so new to cgal i have no idea if this is by design or a bug ;) 

Cheers and thanks for listening!

**Comment by MaelRL** [comment]
This is by design: weight 0 means the skeleton graph is no longer planar, and the skeleton cannot be represented with the current data structures. It is the same with mixing negative and positive weights: you can have all positive, or all negative, but no a bit of both.

But you can work around it:  in the straight skeleton extrusion code, a weight `0` is actually allowed as we really wanted to handle vertically slanted roof sections. The way it is implemented is in fact just a small hack: a tiny, non-zero weight is assigned to the edge, the skeleton is computed with that weight, and the final skeleton is snapped back to vertical.

See:
- https://github.com/CGAL/cgal/blob/master/Straight_skeleton_extrusion_2/include/CGAL/extrude_skeleton.h#L953 (computing a relatively small weight and assigning it to "0" edges)
- https://github.com/CGAL/cgal/blob/master/Straight_skeleton_extrusion_2/include/CGAL/extrude_skeleton.h#L98 (snapping positions).

Maybe it would make sense to do something like this in your setting?

**Comment by pcace** [comment]
or is there an easy way to get contour lines from meshes? 
then i could stick with the mesh (wich is really handy, cause i get the volume too) and generate contour lines every x units?

Thanks again for your very nice and fast help!!

**Comment by MaelRL** [comment]
Do you mean completely independently from Straight skeletons? Something like https://github.com/CGAL/cgal/pull/5895?

**Comment by pcace** [comment]
> Do you mean completely independently from Straight skeletons? Something like #5895?

if i understand #5895 correctly the idea is to make a mesh from isocurve? i mean the other way around - i have a mesh and generate isolines from it. these should be offsets from the original polygon_with_holes i guess?

i guess it would be something like - intersect a plane in z axis with the mesh, and return the outline? 
Maybe too complex... 

I am currently looking into your  snapping-hack. but still quite complex for me ; )

**Comment by MaelRL** [comment]
> if i understand https://github.com/CGAL/cgal/pull/5895 correctly the idea is to make a mesh from isocurve? i mean the other way around - i have a mesh and generate isolines from it. these should be offsets from the original polygon_with_holes i guess?

That pull request is if you have a surface mesh (so, living in 3D), and you want to get an iso-curve from a set of of seeds.

If you mean isocurve from the boundary of 2D polygons, then that's what the offset functions of the straight skeleton package are for: https://doc.cgal.org/5.6-beta1/Straight_skeleton_2/index.html#title11 (you should check this documentation, by the way as it was fully rewritten for the next release, but doc.cgal.org is still at CGAL 5.5).

> I am currently looking into your snapping-hack. but still quite complex for me ; )

Maybe you can just put a very tiny weight for now, and worry about the snapping later.

**Comment by pcace** [comment]
i think sice this https://github.com/CGAL/cgal/issues/7612#issuecomment-1648176800 is not a bug but by design this can be closed here. Thank you so much for you explanations!!
